### PR TITLE
Improve StringBasedDateTime and StringBasedTimeDelta

### PR DESCRIPTION
**Pull request by BenjaminPelletier**
These two types currently have somewhat-opaque and sometimes surprising behavior.  This PR standardizes, documents, and tests their behavior so expectations are clear.  The PR does change undocumented edge case behavior, especially with regard to always producing a timezone-aware datetime in StringBasedDateTime (likely the least-surprising behavior overall).  It also changes behavior to preserve the original data whenever practical by default.

### CODE REVIEW

File: tests/test_string_based_date_time.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    sbdt = StringBasedDateTime(t.isoformat())
+    assert abs(sbdt.datetime - t) <= ZERO_SKEW
+
+    t = arrow.get("1234-01-23T12:34:56.12345678")
+    sbdt = StringBasedDateTime(t)
+    assert "12345678" not in sbdt  # arrow only stores integer microseconds
+
+    with pytest.raises(ValueError):  # unconverted data remains (datetime only parses down to microseconds)
+        _ = datetime.strptime("1234-01-23T12:34:56.12345678", "%Y-%m-%dT%H:%M:%S.%f")
```
Comment (barroco) [inline tests/test_string_based_date_time.py:50]: Out of curiosity, is `_ = ` required ?

File: tests/test_string_based_date_time.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    assert StringBasedDateTime(s) == s
+    assert StringBasedDateTime(s, reformat=True) != s
+
+
+def test_zulu_default():
+    """When a non-string datetime is provided, expect the string representation to use Z as the UTC timezone."""
+
+    assert StringBasedDateTime(datetime.utcnow()).endswith('Z')
+    assert StringBasedDateTime(arrow.utcnow().datetime).endswith('Z')
+    assert StringBasedDateTime(arrow.utcnow()).endswith('Z')
```
Comment (barroco) [inline tests/test_string_based_date_time.py:74]: For completeness, I think a test to check that the Z representation is properly used when a string datetime is provided with reformat.

File: tests/test_string_based_date_time.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    t = arrow.now('US/Pacific')
+    sbdt = StringBasedDateTime(t)
+    assert abs(sbdt.datetime - t) <= ZERO_SKEW
+
+    sbdt = StringBasedDateTime(t.isoformat())
+    assert abs(sbdt.datetime - t) <= ZERO_SKEW
+
+    t = arrow.get("1234-01-23T12:34:56.12345678")
+    sbdt = StringBasedDateTime(t)
+    assert "12345678" not in sbdt  # arrow only stores integer microseconds
```
Comment (barroco) [inline tests/test_string_based_date_time.py:47]: Small suggestion, the current comment gives the impression that the value is cast to microseconds whereas the value is actually rounded to the microseconds.
Example:
```
>>> arrow.get("1234-01-23T12:34:56.12345678").isoformat()
'1234-01-23T12:34:56.123457+00:00'
```

**Review by barroco — APPROVED** [review]
Looks good to me. Few comments / suggestions inline

File: tests/test_string_based_date_time.py (hunk 1 above)
Comment (BenjaminPelletier) [reply tests/test_string_based_date_time.py:50]: No, not required.  I was initially trying to write things that people might attempt and that would usually involve assignment, but assignment isn't part of the mechanism that demonstrates the behavior.  Removed.

File: tests/test_string_based_date_time.py (hunk 2 above)
Comment (BenjaminPelletier) [reply tests/test_string_based_date_time.py:74]: Sounds good; added as elements of test_non_mutation

File: tests/test_string_based_date_time.py (hunk 3 above)
Comment (BenjaminPelletier) [reply tests/test_string_based_date_time.py:47]: Added clarification
